devolving_music/lib/song_scores.py

```diff
@@ -19,28 +19,52 @@ def __init__(self, event: Event):
         self.song_score_dict = ScoreSuite.get_song_scores_dict(event)
         self.comparison_submissions = ScoreSuite.get_event_comparisons(event)
 
-    def get_scores(self) -> "dict[int, ScoreSuite]":
+    def update_scores(self) -> "None":
         # calculates scores for all submissions using current comparisons
+        # returns a dictionary where keys are the song submission id
+        # corresponding to the score suite object
         for compare in self.comparison_submissions:
             song_suite_1 = self.song_score_dict[compare.first_submission.id]
             song_suite_2 = self.song_score_dict[compare.second_submission.id]
 
             SongScores.update_song_rating(compare, song_suite_1, song_suite_2)
 
+    def get_scores_list(self) -> "list[ScoreSuite]":
+        # calculates scores for all submissions using current comparisons
+        # returns list of all score suite objects
+        self.update_scores()
+        return list(self.song_score_dict.values())
+
+    def get_scores_dict(self) -> "dict[int, ScoreSuite]":
+        # calculates scores for all submissions using current comparisons
+        # returns list of all score suite objects
+        self.update_scores()
         return self.song_score_dict
 
     def get_compare_submission_random(self, submission_id) -> SongSubmission:
         key_list = list(self.song_score_dict.keys())
         key_list.remove(submission_id)
-        # once you have a critical number of comparisons then pull from quality
-        # list
-        return self.song_score_dict.get(random.choice(key_list)).song_submission
+        return self.song_score_dict.get(
+            random.choice(key_list)).song_submission
+
+    def get_compare_submission_closest(self, score_suite_obj: "ScoreSuite", information_threshold : int, update_scores = False):
+        # information threshold is the threshhold at which we start to consider score_suites as being able to be close to score_suite_obj
+        # score_suites below this information threshhold are considered to be infinitely far away from score_suite_obj
+        if(update_scores):
+            self.update_scores()
```

Comment:
Still not sure why we'd be updating scores here rather than in some outer context. It doesn't seem related to this method.

I think there's also a good case to be made for taking it out of `get_scores_list` and `get_scores_dict`.

Actually, upon closer inspection, `update_scores` doesn't appear to be idempotent, so calling it all over the place might not only be wasteful but actively cause bugs. Why not just put it in the constructor?

Reply:
It won't cause bugs because the comparison is never changing and that is what the update scores calculation is based on . But you're right I'll put it in the constructor.